wallet-new/integration/Functions.hs

```diff
@@ -35,454 +37,455 @@ import qualified Pos.Wallet.Web.ClientTypes.Types as V0
     => WalletClient m
-    -> WalletState
     -> Action
-    -> m WalletState
+    -> m ()
 -- Wallets
-runAction wc ws  PostWallet = do
-    newPassword <- liftIO $ generate arbitrary
-    newWall     <- liftIO $ generate $ generateNewWallet newPassword
-    result      <- respToRes $ postWallet wc newWall
+runAction wc action = do
+    log $ "Running: " <> show action
+    case action of
+        PostWallet -> do
+            newPassword <- liftIO $ generate arbitrary
+            newWall     <- liftIO $ generate $ generateNewWallet newPassword
+            log $ "Request: " <> show newWall
+            result      <- respToRes $ postWallet wc newWall
 
-    checkInvariant
-        (walBalance result == minBound)
-        (WalletBalanceNotZero result)
+            checkInvariant
+                (walBalance result == minBound)
+                (WalletBalanceNotZero result)
 
-    -- Modify wallet state accordingly.
-    pure $ ws
-        & wallets    <>~ [result]
-        & walletsPass . at result ?~ newPassword
-        & actionsNum +~ 1
-  where
-    generateNewWallet spendPass =
-        NewWallet
-            <$> arbitrary
-            <*> pure (Just spendPass)
-            <*> arbitrary
-            <*> pure "Wallet"
-            <*> pure CreateWallet
+            -- Modify wallet state accordingly.
+            wallets    <>= [result]
+            walletsPass . at result ?= newPassword
+          where
+            generateNewWallet spendPass =
+                NewWallet
+                    <$> arbitrary
+                    <*> pure (Just spendPass)
+                    <*> arbitrary
+                    <*> pure "Wallet"
+                    <*> pure CreateWallet
 
-runAction wc ws GetWallets   = do
-    -- We choose from the existing wallets.
-    result  <-  respToRes $ getWallets wc
+        GetWallets -> do
+            localWallets <- use wallets
+            -- We choose from the existing wallets.
+            result  <-  respToRes $ getWallets wc
 
-    checkInvariant
-        (length result == length (ws ^. wallets))
-        (LocalWalletsDiffers result)
+            checkInvariant
+                (length result == length localWallets)
+                (LocalWalletsDiffers result localWallets)
 
-    -- No modification required.
-    pure $ ws
-        & actionsNum +~ 1
+            -- No modification required.
 
-runAction wc ws GetWallet    = do
-    -- We choose from the existing wallets.
-    wallet  <-  pickRandomElement (ws ^. wallets)
-    result  <-  respToRes $ getWallet wc (walId wallet)
+        GetWallet -> do
+            ws <- get
 
-    checkInvariant
-        (walBalance result == minBound)
-        (LocalWalletDiffers result)
+            let localWallets = ws ^. wallets
 
-    -- No modification required.
-    pure $ ws
-        & actionsNum +~ 1
+            -- The precondition is that we need to have wallets.
+            guard (not (null localWallets))
```

Comment:
I believe `pickRandomElement` will be the guard as well - so maybe this is not needed